<table align="left" width=100%>
    <tr>
        <td width="20%">
            <img src="GL-2.png">
        </td>
        <td>
            <div align="center">
                <font color="#21618C" size=8px>
                  <b> Faculty Notebook <br> (Day 3) </b>
                </font>
            </div>
        </td>
    </tr>
</table>

## Table of Content

1. **[Import Libraries](#lib)**
2. **[Parameter Estimation](#est)**
    - 2.1 - **[Point Estimation](#pt)**
        - 2.1.1 - **[Sampling Error](#err)**
    - 2.2 - **[Interval Estimation for Mean](#int)**
        - 2.2.1 - **[Large Sample Size](#large)**   
        - 2.2.2 - **[Small Sample Size](#small)**
    - 2.3 - **[Interval Estimation for Proportion](#prop)**
3. **[Test of Hypothesis](#defn)**
    - 3.1 - **[Types of Hypothesis](#types)**
4. **[Types of Test](#test_type)**
    - 4.1 - **[Two Tailed Test](#2tailed)**
    - 4.2 - **[One Tailed Test](#1tailed)**
5. **[Hypothesis Tests with Z Statistic](#eg)**
6. **[Error](#error)**

<a id="lib"></a>
# 1. Import Libraries

**Let us import the required libraries.**

In [5]:
# import 'pandas'
import pandas as pd

# import 'numpy'
import numpy as np

# import subpackage of matplotlib
import matplotlib.pyplot as plt

# import 'seaborn'
import seaborn as sns

# to suppress warnings
from warnings import filterwarnings
filterwarnings('ignore')

# import 'random' to generate random sample
import random


# import 'stats' package from scipy library
from scipy import stats

In [ ]:
# set the plot size using 'rcParams'
# once the plot size is set using 'rcParams', it sets the size of all the forthcoming plots in the file
# pass width and height in inches to 'figure.figsize'
plt.rcParams['figure.figsize'] = [15,8]

<a id="est"></a>
# 2. Parameter Estimation

The value associated with the characteristic of the population is known as a `parameter` and the characteristic of the sample is described by a `statistic`.
Usually, the population parameters are not known in most of the real-life problems. Thus, we consider a subset of the population (sample) to estimate the population parameter using a sample statistic.

`Point estimation` and `Interval estimation` are two of the methods to estimate the population parameter.

<a id="pt"></a>
## 2.1 Point Estimation

This method considers a single value (sample statistic) as estimate of the population parameter.

Let $X_{1}, X_{2}, X_{3},..., X_{n}$ be the random sample drawn from a population with mean $\mu$ and standard deviation $\sigma$. <br>
The point estimation method estimates the population mean, $\mu = \overline{X}$, where $\overline{X}$ is the sample mean and population standard deviation, $\sigma = s$, where $s$ is the standard deviation of the sample (`Standard Error`).

### Example:

#### 1. Consider the data of grade points for 35 students in a data science course. Select grades of 20 students randomly from the data and find the point estimate for the population mean.

     Grades: [59.1, 65.0, 75.8, 79.2, 95.0, 99.8, 89.1, 65.2, 41.9, 55.2, 94.8, 84.1, 83.2, 74.0, 75.5, 76.2, 79.1, 80.1,
              92.1, 74.2, 59.2, 64.0, 75, 78.2, 95.6, 97.8, 89.5, 64.2, 41.8, 57.2, 85, 91.4, 81.8, 74.6, 90]

In [10]:
# given population
# draw more samples from the data ~ sampling with replacement
# In Random Forest ~ Bootstrap Aggregation

# given population
grades = [59.1, 65.0, 75.8, 79.2, 95.0, 99.8, 89.1, 65.2, 41.9, 55.2, 94.8, 84.1, 83.2, 74.0, 75.5, 76.2, 79.1, 80.1,
          92.1, 74.2, 59.2, 64.0, 75, 78.2, 95.6, 97.8, 89.5, 64.2, 41.8, 57.2, 85, 91.4, 81.8, 74.6, 90]

import random
print("Pop mean: ", np.mean(grades))
np.mean(random.sample(grades, 20))

random.choices(grades, k=36) # Sampling with replacement

Pop mean:  76.68285714285715


[94.8,
 79.1,
 75.5,
 81.8,
 41.9,
 76.2,
 41.9,
 80.1,
 80.1,
 55.2,
 95.0,
 74.0,
 95.0,
 85,
 99.8,
 74.2,
 90,
 78.2,
 41.8,
 80.1,
 99.8,
 75.8,
 76.2,
 92.1,
 75.8,
 55.2,
 74.0,
 59.2,
 41.9,
 89.5,
 95.0,
 74.0,
 92.1,
 78.2,
 75,
 65.2]

#### 2. A financial firm has created 50 portfolios. From them, a sample of 13 portfolios was selected, out of which 8 were found to be underperforming. Can you estimate the number of underperforming portfolios in the population?

In [11]:
(8/13)*50

30.76923076923077

<a id="err"></a>
### 2.1.1 Sampling Error

Sampling error is considered as the absolute difference between the sample statistic used to estimate the parameter and the corresponding population parameter. Since the entire population is not considered as the sample, the values of mean, median, quantiles, and so on calculated on sample differ from the actual population values.

One can reduce the sampling error either by increasing the sample size or determining the optimal sample size using various methods.

### Example:

#### 1. Consider the data for the number of ice-creams sold per day. An ice-cream vendor collected this data for 90 days and then a sample is drawn (without replacement) containing ice-creams sold for 25 days.

data = [21, 93, 62, 76, 73, 20, 56, 95, 41, 36, 38, 13, 80, 88, 34, 18, 40, 11,
        25, 29, 61, 23, 82, 10, 92, 69, 60, 87, 14, 91, 94, 49, 57, 83, 96, 55,
        79, 52, 59, 39, 58, 17, 19, 98, 15, 54, 48, 46, 72, 45, 65, 28, 37, 30,
        68, 75, 16, 33, 31, 99, 22, 51, 27, 67, 85, 47, 44, 77, 64, 97, 84, 42,
        90, 70, 74, 89, 32, 26, 24, 12, 81, 53, 50, 35, 71, 63, 43, 86, 78, 66]
        
sample = [10, 22, 47, 66, 11, 57, 77, 98, 31, 63, 74, 84, 50, 96, 88, 92, 70, 54, 65, 44, 16, 72, 20, 90, 43]

Comupte the sampling error for mean.

In [12]:
# The absolute value of the difference between the sample mean denoted as x̄,
# and the population mean is denoted by μ, written as |x̄ − μ|, is known as the Sampling Error.

# given population
data = [21, 93, 62, 76, 73, 20, 56, 95, 41, 36, 38, 13, 80, 88, 34, 18, 40, 11, 25, 29, 61, 23, 82, 10, 92, 69, 60, 87, 14, 91,
        94, 49, 57, 83, 96, 55, 79, 52, 59, 39, 58, 17, 19, 98, 15, 54, 48, 46, 72, 45, 65, 28, 37, 30, 68, 75, 16, 33, 31, 99,
        22, 51, 27, 67, 85, 47, 44, 77, 64, 97, 84, 42, 90, 70, 74, 89, 32, 26, 24, 12, 81, 53, 50, 35, 71, 63, 43, 86, 78, 66]

# given sample
sample = [10, 22, 47, 66, 11, 57, 77, 98, 31, 63, 74, 84, 50, 96, 88, 92, 70, 54, 65, 44, 16, 72, 20, 90, 43]

mean_data = np.mean(data)
mean_sample = np.mean(sample)

print("Sampling Error: ", abs(mean_sample - mean_data))


Sampling Error:  3.1000000000000014


<a id="int"></a>
## 2.2 Interval Estimation for Mean

This method considers the range of values in which the population parameter is likely to lie. The confidence interval is an interval that describes the range of values in which the parameter lies with a specific probability. It is given by the formula,<br> <p style='text-indent:20em'> `conf_interval = sample statistic ± margin of error`</p>

The uncertainty of an estimate is described by the `confidence level` which is used to calculate the margin of error.

<a id="large"></a>
### 2.2.1 Large Sample Size

Consider a population with mean $\mu$ and standard deviation $\sigma$. Let us take a sample of `n` observations from the population such that, $n \geq 30$. The central limit theorem states that the sampling distribution of mean follows a normal distribution with mean $\mu$ and standard deviation $\frac{\sigma}{\sqrt(n)}$.

The confidence interval for the population mean with $100(1-\alpha)$% confidence level is given as: $\overline{X} \pm Z_{\frac{\alpha}{2}}\frac{\sigma}{\sqrt{n}}$

Where, <br>
$\overline{X}$: Sample mean<br>
$\alpha$: Level of significance<br>
$\sigma$: Population standard deviation<br>
$n$: Sample size

The quantity $\frac{\sigma}{\sqrt{n}}$ is the standard error of the mean. And the margin of error is given by $Z_{\frac{\alpha}{2}}\frac{\sigma}{\sqrt{n}}$.

If we know the expected margin of error (ME), then we can calculate the required sample size (n) using the formula: $n = (Z_{\frac{\alpha}{2}})^{2}\frac{\sigma^{2}}{ME^{2}}$.

The above equation is valid for any population provided the sample size is sufficiently large (usually $n \geq 30$). Relace $\sigma$ by the standard deviation of the sample ($s$) if the population standard deviation is not known.

The value of $Z_{\frac{\alpha}{2}}$ for different $\alpha$ values can be obtained using the `stats.norm.isf()` from the scipy library.

In [18]:
# Calculating Z Critical

import scipy.stats as stats

conf = 0.99
alpha = 1 - conf

#stats.norm.ppf(conf+alpha/2)
stats.norm.isf(alpha/2) # isf: isotonic survival function
# ppf: percent point function
# sf: survival function


2.5758293035489004

### Book Question from Prem S Mann

In [19]:
xbar = 145
sigma = 35
n = 25

se = sigma/np.sqrt(n)
stats.norm.interval(0.90, loc = xbar, scale = se)

(133.4860246113397, 156.5139753886603)

To calculate the confidence interval with 95% confidence, use the Z-value corresponding to `alpha = 0.05`.

### Example:

#### 1. A random sample of weight (in kg.) for 35 diabetic patients is drawn from the population with a standard deviation of 8 kg. Find the 90% confidence interval for the population mean.

    Weight: [59.1, 65.0, 75.8, 79.2, 95.0, 99.8, 89.1, 65.3, 41.9, 55.2, 94.8, 84.1, 83.2, 74.0, 75.5, 76.2, 79.1, 80.1,
             92.1, 74.2, 59.2, 64.0, 75, 78.2, 95.6, 97.8, 89.5, 64.2, 41.8, 57.2, 85, 91.4, 81.8, 74.6, 90]

In [25]:
weight = [59.1, 65.0, 75.8, 79.2, 95.0, 99.8, 89.1, 65.3, 41.9,
         55.2, 94.8, 84.1, 83.2, 74.0, 75.5, 76.2, 79.1, 80.1,
         92.1, 74.2, 59.2, 64.0, 75, 78.2,
         95.6, 97.8, 89.5, 64.2, 41.8, 57.2, 85, 91.4, 81.8, 74.6, 90]

In [28]:
# Assume the data to be population
sigma = 8
#np.std(weight, ddof = 0) # ddof = 1 represents n-1
n = len(weight)
xbar = np.mean(weight)
print(xbar), print(n)
cf = 0.90

stats.norm.interval(cf, loc = xbar, scale = sigma/np.sqrt(n))

76.68571428571428
35


(74.46146621975642, 78.90996235167215)

#### 2. There are 150 apples on a tree. You randomly choose 40 apples and found that the average weight of apples is 182 grams with a standard deviation of 30 grams. Find the 95% confidence interval for the population mean.

In [32]:
n = 40
xbar = 182
sigma = 30
cf = 0.95

stats.norm.interval(cf, loc = xbar, scale = sigma/np.sqrt(n))

(172.70307451543158, 191.29692548456842)

#### 3. A movie production house needs to estimate the average monthly wage of the technical crew members. The previous data shows that the standard deviation of the wages is 190 dollars. The production team thinks that the estimation of the average wage should not exceed 54 dollars. The team has decided to take a small subset of wages for the estimation. Find a suitable number of wages to be considered to get the estimate with 90% confidence.

#### 4. From a sample of 250 observations, it is found that the average income of a 27 year old Londoner is £45,000 with a sample standard deviation of £4000. Obtain the 95% confidence interval to estimate the average income.

In [37]:
stats.norm.interval(0.95, loc = 45000, scale = 4000/np.sqrt(250))

(44504.16397415635, 45495.83602584365)

### Finding if the Population Mean truly exists in Conf Interval Range

Note: We will perform it using 0.95

In [42]:
# Considering the entire data as Population Data
tips = sns.load_dataset("tips")
sigma = np.std(tips.total_bill, ddof = 0)

# Draw a Sample of 30 Rows
sample = tips.sample(30, random_state = 20)

xbar = np.mean(sample.total_bill)
n = sample.shape[0]

print("Population Mean: ", tips.total_bill.mean())
stats.norm.interval(0.95, loc = xbar, scale = sigma/np.sqrt(n))

Population Mean:  19.78594262295082


(17.071905962870094, 23.430094037129894)

<a id="small"></a>
### 2.2.2 Small Sample Size

Let us take a sample of `n` observations from the population such that, $n < 30$. Here the standard deviation of the population is unknown. The confidence interval for the population mean with $100(1-\alpha)$% confidence level is given as: $\overline{X} \pm t_{\frac{\alpha}{2}, n-1}\frac{s}{\sqrt{n}}$

Where, <br>
$\overline{X}$: Sample mean<br>
$\alpha$: Level of significance<br>
$s$: Sample standard deviation<br>
$n-1$: degrees of freedom

The ratio $\frac{s}{\sqrt{n}}$ is the estimate of the standard error of the mean. And $t_{\frac{\alpha}{2}, n-1}\frac{s}{\sqrt{n}}$ is the margin of error for the estimate.

The value of $t_{\frac{\alpha}{2}, n-1}$ for different $\alpha$ values can be obtained using the `stats.t.isf()` from the scipy library.  

### Example:

Dr. Moore wanted to estimate the mean cholesterol level for all adult men living in Hartford.

He took a sample of 25 adult men from Hartford and found that the mean cholesterol level for this sample is 186 mg/dL with a standard deviation of 12 mg/dL.

Assume that the cholesterol levels for all adult men in Hartford are (approximately) normally distributed.

Construct a 95% confidence interval for the population mean .

In [43]:
n = 25
xbar = 186
samplesd = 12
cf = 0.95

# Since we have been given the sample SD, we use t-distt to calc conf
stats.t.interval(cf, df = n-1, loc = xbar, scale = samplesd/np.sqrt(n))

(181.04664345209275, 190.95335654790725)

<a id="prop"></a>
## 2.3 Interval Estimation for Proportion - Additional Topic

Consider a population in which each observation is either a success or a failure. The population proportion is denoted by `P` which the ratio of the number of successes to the size of the population.

The confidence interval for the population proportion with $100(1-\alpha)$% confidence level is given as: $p \pm Z_{\frac{\alpha}{2}}\sqrt{\frac{p(1 - p)}{n}}$

Where, <br>
$p$: Sample proportion<br>
$\alpha$: Level of significance<br>
$n$: Sample size

The quantity $Z_{\frac{\alpha}{2}}\sqrt{\frac{p(1 - p)}{n}}$ is the margin of error.

### Example:

#### 1. A financial firm has created 50 portfolios. From them, a sample of 13 portfolios was selected, out of which 8 were found to be underperforming. Construct a 99% confidence interval to estimate the population proportion.

In [ ]:
# Confidence Interval for Proportions


<a id="defn"></a>
# 3. Test of Hypothesis

Hypothesis testing is the process of evaluating the validity of claims made about the population using the sample data obtained from the population. A statistical test is a rule used to decide whether to reject or retain the claim.

**Examples of hypothesis:**

        1. One can get 'A' grade if the attendance in the class is more than 75%.
        2. A probiotic drink can improve the immunity of a person.

<a id="types"></a>
## 3.1 Types of Hypothesis

`Null Hypothesis`: The null hypothesis is the statistical hypothesis suggesting 'no difference' between the population parameter and a specific value
. It is denoted as H<sub>0</sub>.

`Alternative Hypothesis`: It is the hypothesis that is tested against the null hypothesis and states the existence of a difference between the parameter and a specific value. It is denoted by H<sub>a</sub> or H<sub>1</sub>.

The claim is usually the alternative hypothesis H<sub>1</sub> also known as research hypothesis. To test the claim we need to gather evidence (data ) and find the likelihood of the data under the assumption H<sub>0</sub> being true.

#### Example:

A company that produces tennis balls claimed that the diameter of a tennis ball is at least 2.625 inches on average. On the other hand, a professional tennis coach claimed that the diameter of a ball is less than what the company has claimed. To test the claim of the coach, a statistical test can be performed considering the hypothesis:

                    Null Hypothesis: Average diameter ≥ 2.625
                    Alternative Hypothesis: Average diameter < 2.625

<a id="test_type"></a>
# 4. Types of Test

The hypothesis test is used to validate the claims about the population. The types of tests are based on the nature of the alternative hypothesis.

<a id="2tailed"></a>
## 4.1 Two Tailed Test

Two tailed test considers the value of the population parameter is less than or greater than (i.e. not equal) a specific value. <br>
If we test the population mean ($\mu$) with a specific value ($\mu_{0}$) the null hypothesis is: $H_{0}: \mu = \mu_{0}$.

The alternative hypothesis for the two tailed test is given as: $H_{1}: \mu \neq \mu_{0}$

#### Example:

A company that produces tennis balls claimed that the diameter of a tennis ball is 2.625 inches on average. To test the company's claim, a statistical test can be performed considering the hypothesis:

                    Null Hypothesis: Average diameter = 2.625
                    Alternative Hypothesis: Average diameter ≠ 2.625

<a id="1tailed"></a>
## 4.2 One Tailed Test

One tailed test considers the value of the population parameter is less than or greater than (but not both) a specific value. <br>
If we test the population mean ($\mu$) with a specific value ($\mu_{0}$) the null hypothesis is: $H_{0}: \mu \leq \mu_{0}$ and the alternative hypothesis is $H_{1}: \mu > \mu_{0}$, the one tailed test is also known as a `right-tailed test`.

If we test the population mean ($\mu$) with a specific value ($\mu_{0}$) the null hypothesis is: $H_{0}: \mu \geq \mu_{0}$ and the alternative hypothesis is $H_{1}: \mu < \mu_{0}$, the one tailed test is also known as a `left-tailed test`.


### Example:

**1.** The company's annual quality report of machines states that a lathe machine works efficiently at most for 8 months on average after the servicing. The production manager claims that after the special tuxan servicing, the machine works efficiently for more than 8 months. To test the claim of production manager consider the hypothesis:

                    Null Hypothesis: Machine efficiency ≤ 8 months
                    Alternative Hypothesis: Machine efficiency > 8 months

This is the example of a **right-tailed test**.

**2.** A railway authority claims that all the trains on the Chicago-Seattle route run with a speed of at least 54 mph on average. A customer forum declares that there are various records from passengers claiming that the speed of the train is less than what railway has claimed. In this scenario, a statistical test can be performed to test the claim of customer forum considering the hypothesis:

                    Null Hypothesis: Speed ≥ 56 mph
                    Alternative Hypothesis: Speed < 56 mph

This is the example of a **left-tailed test**.

<a id="eg"></a>
# 5. Hypothesis Tests with Z Statistic

Let us perform one sample Z test for the population mean. We compare the population mean with a specific value. The sample is assumed to be taken from a population following a normal distribution.

To check the normality of the data, a test for normality is used. The `Shapiro-Wilk Test` is one of the methods used to check the normality. The hypothesis of the test is given as:
<p style='text-indent:25em'> <strong> H<sub>0</sub>:  The data is normally distributed </strong> </p>
<p style='text-indent:25em'> <strong> H<sub>1</sub>:  The data is not normally distributed </strong> </p>

The `shapiro()` from scipy library performs a Shapiro-Wilk normality test.

The null and alternative hypothesis of Z-test is given as:
<p style='text-indent:25em'> <strong> $H_{0}: \mu = \mu_{0}$ or $\mu \geq \mu_{0}$ or $\mu \leq \mu_{0}$</strong></p>
<p style='text-indent:25em'> <strong> $H_{1}: \mu \neq \mu_{0}$ or $\mu < \mu_{0}$ or $\mu > \mu_{0}$</strong></p>

Consider a normal population with standard deviation $\sigma$. Let us take a sample of size n, such that (n > 30).
The test statistic for one sample Z-test is given as:
<p style='text-indent:25em'> <strong> $Z = \frac{\overline{X} -  \mu}{\frac{\sigma}{\sqrt(n)}}$</strong></p>

Where, <br>
$\overline{X}$: Sample mean<br>
$\mu$: Specified mean<br>
$\sigma$: Population standard deviation<br>
$n$: Sample size

Under $H_{0}$ the test statistic follows a standard normal distribution.

If $\sigma$ is unknown, use the sample standard deviation (s) instead of $\sigma$ to calculate the test statistic.

### Example:

#### 1. A car manufacturing company claims that the mileage of their new car is 25 kmph with a standard deviation of 2.5 kmph. A random sample of 45 cars was drawn and recorded their mileage as per the standard procedure. From the sample, the mean mileage was seen to be 24 kmph. Is this evidence to claim that the mean mileage is different from 25kmph? (assume the normality of the data) Use α = 0.01.

The null and alternative hypothesis is:

H<sub>0</sub>: $\mu = 25 $<br>
H<sub>1</sub>: $\mu ≠ 25 $

Here ⍺ = 0.01, for a two-tailed test calculate the critical z-value.

i.e. if z is less than -2.58 or greater than 2.58 then we reject the null hypothesis.

Here the z score is less than -2.58, p-value is less than 0.01, also the confidence interval do not contain the value in the null hypothesis (i.e. 25), thus we reject the null hypothesis and conclude that there is enough evidence to claim that the mean mileage is different from 25kmph.

#### 2. The average calories in a slice bread of the brand 'Alphas' are 82 with a standard deviation of 15. An experiment is conducted to test the claim of the dietitians that the calories in a slice of bread are not as per the manufacturer's specification. A sample of 40 slices of bread is taken and the mean calories recorded are 95. Test the claim of dietitians with ⍺ value (significance level) as 0.05. (assume the normality of the data).

The null and alternative hypothesis is:

H<sub>0</sub>: $\mu = 82 $<br>
H<sub>1</sub>: $\mu ≠ 82 $

Here ⍺ = 0.05, for a two-tailed test calculate the critical z-value.

#### 3. An e-commerce company claims that the mean delivery time of food items on its website in NYC is 60 minutes with a standard deviation of 30 minutes. A random sample of 45 customers ordered from the website, and the average time for delivery was found to be 75 minutes. Is this enough evidence to claim that the average time to get items delivered is more than 60 minutes. (assume the normality of the data). Test the with α = 0.05.

The null and alternative hypothesis is:

H<sub>0</sub>: $\mu \leq  60 $<br>
H<sub>1</sub>: $\mu > 60 $

Here ⍺ = 0.05, for a one-tailed test calculate the critical z-value.

i.e. if z is greater than 1.64 then we reject the null hypothesis.

Here the z score is greater than 1.64, the p-value is less than 0.05 thus we reject the null hypothesis and conclude that the average time to get items delivered is more than 60 minutes.

<a id="error"></a>
# 6. Errors in Hypothesis Testing

Under the hytpothesis testing framework there can be two decisions; either we 'reject' the null hypotyhesis H<sub>0</sub> or 'not reject' the null hypothesis H<sub>0</sub>. Errors can still occur in a hypothesis testing scenario because the decision to reject or not reject the null hypothesis is made on the basis of data taken from a sample. Two types of errors are `Type I` and `Type II` error.

### Type I Error

This kind of error occurs when we reject the null hypothesis even if it is true. It is equivalent to a `false positive` conclusion. The maximum probability of committing a type I error is given by the value of $\alpha$, level of significance. So when $\alpha$ = 0.05 there is a 5% chance of rejecting a true null hypothesis.

### Type II Error

This kind of error occurs when we fail to reject the null hypothesis even if it is wrong. It is equivalent to a `false negative` conclusion. The probability of type II error is given by the value of $\beta$. The value of $\beta$ cannot be easily calculated. But $\alpha$ and $\beta$ are thus related that decreasing one increases the other.